### PR TITLE
Support VSO package manager registry

### PR DESCRIPTION
**Pull request by lyweiwei**
This is for a pre-review for the fix of #2505,
Will add test cases if the maintainer team agrees the basic idea.

### Root cause
As described in #2505, for VSO package manager registries the package tarball's URL is not prefixed with the registry URL.

### Fix
When downloading tarball, calculate the auth method with the original package nam instead of the tarball URL, if the URL is resolved from an NPM registry).

### CODE REVIEW

**Comment by bestander** [comment]
Tests are broken.
Also very much appreciated would be new tests added

**Comment by lyweiwei** [comment]
@bestander I just pushed a fix for the failed test cases. But it seems some of the cases timeout due to network issues. Is this a know issue?

> Summary of all failing tests
> FAIL  __tests__/commands/install/integration.js (205.516s)
>  ● install incompatible optional dependency should still install shared child dependencies
>
>    Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.
>      
>      at ontimeout (timers.js:365:14)
>      at tryOnTimeout (timers.js:237:5)
>      at Timer.listOnTimeout (timers.js:207:5)
>
> FAIL  __tests__/package-resolver.js (88.384s)
>  ● resolve react-native
>
>    Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.
>      
>      at ontimeout (timers.js:365:14)
>      at tryOnTimeout (timers.js:237:5)
>      at Timer.listOnTimeout (timers.js:207:5)

These test cases passed on Travis.CI

**Comment by lyweiwei** [comment]
BTW, can we do a retry on the failed environments?

**Comment by bestander** [comment]
@lyweiwei, yeah I think 3 retries could be reasonable.
Most of the network tests are actually mocked with checked in caches, so CI is not that unstable.
Probably your commit happened during some bad time.

**Comment by bestander** [comment]
Yeah, I think the logic is reasonable.
Please go on with some tests and let's merge it once we are sure we don't break any existing installs

**Comment by zumwald** [comment]
bump @bestander @lyweiwei what's the path forward for vsts support?

**Comment by bestander** [comment]
@zumwald, this PR is quite good, it just needs a test and a rebase.
If you want to take it over, send a new PR with required changes and we'll merge it in.

**Comment by lyweiwei** [comment]
I just pushed a new iteration
* merge the latest changes from master
* wait longer for the package-resolver test cases (6s->10s)

**Comment by lyweiwei** [comment]
@bestander The PR passed the test this time :) sorry for the delay.

File: src/registries/is-request-to-registry.js [hunk 1]
```diff
@@ -2,18 +2,25 @@
 
 import url from 'url';
 
+const SUFFIX_VISUALSTUDIO = '.pkgs.visualstudio.com';
```
Comment (bestander) [inline src/registries/is-request-to-registry.js]: I think I missed that detail earlier.
It looks hard to scale a maintain a solution where every registry has to be hardcoded in Yarn.
Can this setting be moved to .yarnrc for each registry to configure?

File: __tests__/package-resolver.js [hunk 2]
```diff
@@ -10,7 +10,7 @@ import * as fs from '../src/util/fs.js';
 import * as constants from '../src/constants.js';
 import inquirer from 'inquirer';
 
-jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;
+jasmine.DEFAULT_TIMEOUT_INTERVAL = 100000;
```
Comment (bestander) [inline __tests__/package-resolver.js]: Does it mean that requests get slower after this change?

Comment (bestander) [inline __tests__/package-resolver.js]: Also can you add a test here for the supported case?

File: src/registries/npm-registry.js [hunk 3]
```diff
@@ -50,16 +52,14 @@ export default class NpmRegistry extends Registry {
...
 
-  request(pathname: string, opts?: RegistryRequestOptions = {}): Promise<*> {
-    const registry = addSuffix(this.getRegistry(pathname), '/');
+  request(pathname: string, opts?: RegistryRequestOptions = {}, packageName: ?string): Promise<*> {
+    const registry = this.getRegistry(packageName || pathname);
     const requestUrl = url.resolve(registry, pathname);
-    const alwaysAuth = this.getScopedOption(registry.replace(/^https?:/, ''), 'always-auth')
-      || this.getOption('always-auth')
-      || removePrefix(requestUrl, registry)[0] === '@';
+    const alwaysAuth = this.getRegistryOrGlobalOption(registry, 'always-auth');
```
Comment (restrry) [inline src/registries/npm-registry.js]: @bestander does it make sense to force adding auth key to resoled url?
I spent a day some months ago when tried to fix [similar problem](https://github.com/yarnpkg/yarn/issues/521#issuecomment-277500043) with our artifactory.

**Comment by kenotron** [comment]
@lyweiwei Can you remove the VSTS specifics here and make Yarn work against VSO branch? Or did you find a different work around?

**Comment by lumaxis** [comment]
@lyweiwei @bestander I tried picking continuing this in #3231. Would very much appreciate a review! 🙂

**Comment by bestander** [comment]
This one got replaced with #3231

**Pull request closed without merging**